xcode/Subconscious/Shared/Models/MemoAddress.swift

```diff
@@ -9,83 +9,81 @@ import Foundation
 
 enum MemoAddress: Hashable, Codable {
     case local(Slug)
-    case `public`(Slug)
+    case `public`(Slashlink)
     
     var slug: Slug {
         switch self {
         case .local(let slug):
             return slug
-        case .public(let slug):
-            return slug
+        case .public(let slashlink):
+            return slashlink.toSlug()
         }
     }
 }
 
-extension MemoAddress: LosslessStringConvertible, Identifiable {
+extension MemoAddress: LosslessStringConvertible, Identifiable, Comparable {
     /// An address is considered to be any two strings separated by `::`
     private static let addressRegex = /(.+)::(.+)/
     
     /// Initialize from a string
     init?(_ description: String) {
-        guard
-            let match = try? Self.addressRegex.wholeMatch(in: description)
-        else {
+        guard let match = try? Self.addressRegex.wholeMatch(
+            in: description
+        ) else {
             return nil
         }
         guard let audience = Audience(rawValue: String(match.1)) else {
             return nil
         }
-        guard let slug = Slug(String(match.2)) else {
-            return nil
-        }
         switch audience {
         case .local:
-            self = .local(slug)
+            guard let slashlink = Slashlink(String(match.2)) else {
+                return nil
+            }
+            // Local addresses must not include petnames
+            guard slashlink.petnamePart == nil else {
+                return nil
+            }
+            self = .local(slashlink.toSlug())
         case .public:
-            self = .public(slug)
+            guard let slashlink = Slashlink(String(match.2)) else {
+                return nil
+            }
+            self = .public(slashlink)
         }
     }
     
     var description: String {
         switch self {
         case .local(let slug):
-            return "local::\(slug)"
-        case .public(let slug):
-            return "public::\(slug)"
+            return "local::\(slug.toSlashlink())"
```

Comment:
Local MemoAddresses are also serialized as slashlinks, but we can guarantee that they are always local because they are derived from a slug.